src/test/shell/bazel/cc_runfiles_test.sh

```diff
@@ -0,0 +1,366 @@
+  deps = [":library"],
+)
+
+cc_test(
+  name = "feature_test",
+  srcs = ["feature_test.cpp"],
+  deps = ["@bazel_tools//tools/cpp/runfiles"],
+)
+EOF
+
+  cat > pkg/library.cpp <<'EOF'
+#include "library.h"
+#include <iostream>
+void print_repo_name() {
+  std::cout << "in " << __FILE__ << ": '" << BAZEL_CURRENT_REPOSITORY << "'" << std::endl;
+}
+EOF
+
+  cat > pkg/library.h <<'EOF'
+#include <iostream>
+#include "tools/cpp/runfiles/runfiles.h"
+#include "pkg/gen_library.h"
+#include "gen_pkg/gen_library.h"
+
+using bazel::tools::cpp::runfiles::Runfiles;
+
+inline void print_repo_name() {
+  std::cout << "in " << __FILE__ << ": '" << Runfiles::CurrentRepository() << "'" << std::endl;
+  gen_print_repo_name();
+  other_repo_gen_print_repo_name();
+}
+EOF
+
+  cat > pkg/binary.cpp <<'EOF'
+#include <iostream>
+#include "library.h"
+int main() {
+  std::cout << "in " << __FILE__ << ": '" << BAZEL_CURRENT_REPOSITORY << "'" << std::endl;
+  print_repo_name();
+}
+EOF
+
+  cat > pkg/test.cpp <<'EOF'
+#include <iostream>
+#include "library.h"
+int main() {
+  std::cout << "in " << __FILE__ << ": '" << BAZEL_CURRENT_REPOSITORY << "'" << std::endl;
+  print_repo_name();
+}
+EOF
+
+  cat > pkg/feature_test.cpp <<'EOF'
+#include "tools/cpp/runfiles/runfiles.h"
+
+int main() {
+#if defined(BAZEL_TOOLS_CPP_RUNFILES_HAS_BUILTIN_FILE)
+  return 0;
+#else
+  return 1;
+#endif
+}
+EOF
+
+  mkdir -p other_repo
+  touch other_repo/WORKSPACE
+
+  mkdir -p other_repo/pkg
+  cat > other_repo/pkg/BUILD.bazel <<'EOF'
+cc_binary(
+  name = "binary",
+  srcs = ["binary.cpp"],
+  deps = ["@//pkg:library"],
+)
+
+cc_test(
+  name = "test",
+  srcs = ["test.cpp"],
+  deps = ["@//pkg:library"],
+)
+EOF
+
+  cat > other_repo/pkg/binary.cpp <<'EOF'
+#include <iostream>
+#include "pkg/library.h"
+int main() {
+  std::cout << "in " << __FILE__ << ": '" << BAZEL_CURRENT_REPOSITORY << "'" << std::endl;
+  print_repo_name();
+}
+EOF
+
+  cat > other_repo/pkg/test.cpp <<'EOF'
+#include <iostream>
+#include "pkg/library.h"
+int main() {
+  std::cout << "in " << __FILE__ << ": '" << BAZEL_CURRENT_REPOSITORY << "'" << std::endl;
+  print_repo_name();
+}
+EOF
+
+  mkdir -p other_repo/gen_pkg
+  cat > other_repo/gen_pkg/BUILD.bazel <<'EOF'
+genrule(
+  name = "gen_library_hdr",
+  outs = ["gen_library.h"],
+  cmd = """
+cat >$@ <<EOF_
+#include <iostream>
+
+#include "tools/cpp/runfiles/runfiles.h"
+
+using bazel::tools::cpp::runfiles::Runfiles;
+
+inline void other_repo_gen_print_repo_name() {
+  std::cout << "in " << __FILE__ << ": '" << Runfiles::CurrentRepository() << "'" << std::endl;
+}
+EOF_
+"""
+)
+
+cc_library(
+  name = "gen_library",
+  hdrs = ["gen_library.h"],
+  deps = ["@bazel_tools//tools/cpp/runfiles"],
+  visibility = ["//visibility:public"],
+)
+EOF
+
+  bazel build pkg:feature_test &>"$TEST_log" || fail "Build should succeed"
+  if ! bazel test pkg:feature_test &>"$TEST_log"; then
+    echo "test skipped because __builtin_FILE is not supported" >&2
```

Comment:
Do you know which CI runners actually end up running this test? I'm still a bit concerned that it's not running on at least one runner per OS